csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,552 @@ bool predicateAtEnd(ForLoop* loop) {
+// propagation and reached domain. We already proved that [2] is linear in
+// domain. In domain, the domains on the right of the 4 are 8, 8, 8, so the
+// stride is 8*8*8 = 512.
+namespace {
+
+// From the above example, we can see that how linear_g lives in domain could be
+// complicated. It can be, for example:
+//   1. linear_g is equivalent to a single ValGroup in domain.
+//   2. linear_g is the inner of a ValGroup in domain.
+//   3. linear_g is the outer of a ValGroup in domain.
+//   4. linear_g is the middle of a ValGroup in domain, where on the right,
+//      there is a 2.
+//   5. linear_g is split as g1, g2, g3 in domain.
+//   6. linear_g is split as the inner 2 of g1, g2, and the outer 4 of g3.
+//
+// We use a dynamic type to be able to represent all these cases.
+template <typename AbstractValGroup>
+struct PartOf {
+  std::shared_ptr<AbstractValGroup> group;
+  Val* inner_extent = nullptr;
+  Val* selected_extent = nullptr;
+};
+
+using AbstractValGroup = dynamic_type::DynamicType<
+    dynamic_type::Containers<
+        std::deque, // a composition of AbstractValGroups
+        PartOf // part of a AbstractValGroup
+        >,
+    ValGroup // a whole ValGroup
+    >;
+
+// Utilities to print AbstractValGroup.
+std::string print(const AbstractValGroup& group);
+
+std::string print(const ValGroup& group) {
+  return group->toString();
+}
+
+std::string print(const PartOf<AbstractValGroup>& part) {
+  auto str_or_null = [](Val* val) {
+    return val == nullptr ? "nullptr" : val->toInlineString();
+  };
+  return "PartOf(group=" + print(*part.group) +
+      ", inner_extent=" + str_or_null(part.inner_extent) +
+      ", selected_extent=" + str_or_null(part.selected_extent) + ")";
+}
+
+std::string print(const std::deque<AbstractValGroup>& vec) {
+  std::stringstream ss;
+  ss << "[";
+  for (const auto& g : vec) {
+    ss << print(g) << ", ";
+  }
+  ss << "]";
+  return ss.str();
+}
+
+std::string print(const std::monostate&) {
+  return "std::monostate";
+}
+
+std::string print(const AbstractValGroup& group) {
+  return AbstractValGroup::dispatch(
+      [&](const auto& group) { return print(group); }, group);
+}
+
+// Utilities to check if AbstractValGroup is related to ValGroup.
+bool related(const AbstractValGroup& current, const ValGroup& to);
+
+bool related(const ValGroup& current, const ValGroup& to) {
+  return current == to;
+}
+
+bool related(const PartOf<AbstractValGroup>& current, const ValGroup& to) {
+  return related(*current.group, to);
+}
+
+bool related(const std::deque<AbstractValGroup>& current, const ValGroup& to) {
+  return std::any_of(current.begin(), current.end(), [&](const auto& g) {
+    return related(g, to);
+  });
+}
+
+bool related(const std::monostate& current, const ValGroup& to) {
+  return false;
+}
+
+bool related(const AbstractValGroup& current, const ValGroup& to) {
+  return AbstractValGroup::dispatch(
+      [&](const auto& current) { return related(current, to); }, current);
+}
```

Comment:
Did you need to forward-declare this in order for this dispatch call to compile?

Reply:
Forward declare is not for this function body to compile, but instead, is for the definition of
```C++
bool related(const PartOf<AbstractValGroup>& current, const ValGroup& to)
```
and
```C++
bool related(const std::deque<AbstractValGroup>& current, const ValGroup& to)
```
to compile because it uses `bool related(const AbstractValGroup& current, const ValGroup& to)` in its function body.

Reply:
in this context, the type of `current` will be one of `const PartOf<AbstractValGroup>&`, or `const std::deque<AbstractValGroup>&`, or `std::monostate`, so this `related(current, to)` will use the above specialized versions, instead of the abstract version `bool related(const AbstractValGroup& current, const ValGroup& to)`.

Reply:
Got it. Thanks.